### PR TITLE
Add favorite buildings

### LINKED ISSUE #1960: Add Favorites for Building Hours
This would be a nice feature to have, and has been requested by a user.  This could just be a button on the detail view.  Such a button would mark a building as a favorite, and such favorited buildings would be shown at the top of the building hours list view.

### PR DESCRIPTION
**Pull request by hannesmcman**
This should resolve #1960. I added a new component for the favorite button and I don't know if I followed best practices with storing which buildings were favorited, but I learned a lot in the process! I wasn't able to get the app running on an android simulator...kept getting this weird bug about multiple dex files.

### CODE REVIEW

**Review by hawkrives — COMMENTED** [review]
I'll review in a few days with some actual comments, but thanks for the PR! I know our contributing guidelines aren't very useful for figuring out how the app works. 

My first comment: you might want to look at lib/storage.js and create a pair of functions to store the set of favorited buildings so that they're persisted across app restarts.

**Comment by hawkrives** [comment]
@hannesmcman Would you mind running "yarn prettier" on your branch and pushing again, please?

**Comment by drewvolz** [comment]
Hey @hannesmcman, thanks! I’ll also get around to running this on my end in a few days. It is great to see a pull request from you!

**Comment by drewvolz** [comment]
On another note, I feel as if the “multiple dex file” error could be caused by trying to load a copy of AAO on an android sim/device where it’s already installed. Just a hunch.

**Comment by hawkrives** [comment]
Thanks! I think the build will still fail, but that's on me to fix – one of the other PRs I have open – currently CI attempts to sign every built app with release credentials, which I need to disable so that other PRs like this one can pass.

**Comment by hannesmcman** [comment]
@hawkrives I'll take a look at the lib/storage.js. I wasn't sure about how to make the storage persist, so thanks for pointing me in the right direction. No worries on the build failure, should I fix the code climate issues though?

**Comment by hannesmcman** [comment]
@drewvolz As for the "multiple dex files" error..I downloaded a whole new emulator to try and get it to work and the "npm run android" command keeps spitting out this error..I've no idea why.

**Comment by hawkrives** [comment]
Yeah, you should be able to get CodeClimate to pass. I'd recommend trying to join the two if-branches of FavoriteButton into a single branch as much as possible - probably using ternaries to change the icon name, since AFAICT that's all that changes between the two branches. 

For expedience, could you copy/paste the error message you get when running the emulator on your machine into a code block? Also, are you running it on Windows, macOS, or Linux?

**Comment by hawkrives** [comment]
(Expedience because we might possibly be able to debug it without pulling out our laptops if we see the exact error. Maybe.)

**Comment by hawkrives** [comment]
another hint: you might look through how we handle the homescreen sorting; I think that's the closest analogue to what you're trying to do.

**Comment by hannesmcman** [comment]
@hawkrives Here's the error I get when I try to run "npm run android":

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':app:transformClassesWithDexForDebug'.
> com.android.build.api.transform.TransformException: com.android.ide.common.process.ProcessException: java.util.concurrent.ExecutionException: com.android.dex.DexException: Multiple dex files define Landroid/support/v4/accessibilityservice/AccessibilityServiceInfoCompat$AccessibilityServiceInfoVersionImpl;
```

Also, for some reason when I run "npm run check" I get this error:

```
> AllAboutOlaf@2.4.3-pre test /Users/Hannes/AAO-React-Native
> jest

Error: Cannot find module '/Users/Hannes/AAO-React-Native/node_modules/jest-cli'
    at Function.Module._resolveFilename (module.js:489:15)
    at Function.Module._load (module.js:439:25)
    at Module.require (module.js:517:17)
    at require (internal/module.js:11:18)
    at getJest (/Users/Hannes/AAO-React-Native/node_modules/jest/node_modules/jest-cli/build/cli/get_jest.js:23:12)
    at getRunCLIFn (/Users/Hannes/AAO-React-Native/node_modules/jest/node_modules/jest-cli/build/cli/index.js:162:69)
    at Object.<anonymous> (/Users/Hannes/AAO-React-Native/node_modules/jest/node_modules/jest-cli/build/cli/index.js:44:24)
    at Generator.next (<anonymous>)
    at step (/Users/Hannes/AAO-React-Native/node_modules/jest/node_modules/jest-cli/build/cli/index.js:55:2669)
    at /Users/Hannes/AAO-React-Native/node_modules/jest/node_modules/jest-cli/build/cli/index.js:55:2899
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! AllAboutOlaf@2.4.3-pre test: `jest`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the AllAboutOlaf@2.4.3-pre test script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/Hannes/.npm/_logs/2017-12-25T16_30_19_522Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! AllAboutOlaf@2.4.3-pre check: `npm run prettier && npm run flow && npm run validate-data -- --quiet && npm run test`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the AllAboutOlaf@2.4.3-pre check script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/Hannes/.npm/_logs/2017-12-25T16_30_19_557Z-debug.log
```

I am running everything on a mac.

**Comment by drewvolz** [comment]
The project builds and runs on both my iOS and Android emulators. We'll see what we can do to get your Android emulator running.

File: source/flux/parts/buildings.js [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+const SAVE_FAVORITE_BUILDINGS = 'buildings/SAVE_FAVORITE_BUILDINGS'
+
+export async function loadFavoriteBuildings() {
+  let favoriteBuildings = await storage.getFavoriteBuildings()
+  return saveFavoriteBuildings(favoriteBuildings)
+}
+
+type SaveFavoriteBuildingsAction = {
+  type: 'buildings/SAVE_FAVORITE_BUILDINGS',
+  payload: Array<BuildingType>,
```
Comment (hawkrives) [inline source/flux/parts/buildings.js]: It's probably significantly simpler overall to store an array of the building names as strings instead of the actual building information. Also requires less space on the device and doesn't raise so many issues of caching. 

Would you mind working this over to use an array of strings instead of BuildingType, please?

Comment (hawkrives) [reply source/flux/parts/buildings.js]: Well, so much for the "start a review" button… Review in progress!

File: source/lib/storage.js [hunk 2]
```diff
@@ -74,3 +75,13 @@ export function setCredentialsValid(valid: boolean) {
...
   return getItemAsBoolean(credentialsValidKey)
 }
+
+/// MARK: Favorite Buildings
+
+const favoriteBuildingsKey = 'buildings:favorited'
+export function setFavoriteBuildings(buildings: Array<BuildingType>) {
+  return setItem(favoriteBuildingsKey, buildings)
+}
+export function getFavoriteBuildings() {
```
Comment (hawkrives) [inline source/lib/storage.js]: I'd like a return type on this function, please; probably `Promise<Array<BuildingType>>` given this code. It should return the same type as what was stored, essentially.

File: source/views/building-hours/detail/index.js [hunk 3]
```diff
@@ -11,17 +11,45 @@ import moment from 'moment-timezone'
...
+    const building = navigation.state.params.building
+    let favoriteBuildings = navigation.state.params.favoriteBuildings
     return {
-      title: navigation.state.params.building.name,
+      title: building.name,
+      headerRight: (
+        <FavoriteButton
+          favorited={favoriteBuildings.includes(building)}
+          navigation={navigation}
+          onFavorite={() => {
```
Comment (hawkrives) [inline source/views/building-hours/detail/index.js]: I'd like to pull this logic out of here, probably into a function in parts/building-hours.js? Said function would be called as part of the saveBuildingFavorites function.

File: source/views/building-hours/stateful-list.js [hunk 4]
```diff
@@ -27,7 +29,31 @@ const groupBuildings = (buildings: BuildingType[]) => {
   return toPairs(grouped).map(([key, value]) => ({title: key, data: value}))
 }
 
-type Props = TopLevelViewPropsType
+const loadAllBuildings = (
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: I'd like to see this logic incorporated into groupBuildings itself, if you don't mind. I can show you how if you'd like.

File: source/views/building-hours/stateful-list.js [hunk 5]
```diff
@@ -48,16 +74,15 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
     intervalId: 0,
   }
 
   componentWillMount() {
     this.fetchData()
-
     // This updates the screen every ten seconds, so that the building
     // info statuses are updated without needing to leave and come back.
-    this.setState({intervalId: setInterval(this.updateTime, 1000)})
+    this.setState({intervalId: setInterval(this.updateTime, 10000)})
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: This change should be reverted, please.

File: source/views/building-hours/stateful-list.js [hunk 6]
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
     this.setState(() => ({
-      buildings: groupBuildings(buildings),
+      buildings: loadAllBuildings(buildings, this.props.favoriteBuildings),
       now: moment.tz(CENTRAL_TZ),
     }))
   }
 
+  onUpdate = () => {
+    this.fetchData()
+    this.props.onSaveFavorites(this.props.favoriteBuildings)
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: This is the other half of my comment about moving the "make a favorite" logic into the redux file; when you do that, you won't need to save in onUpdate at all, because the update will have happened as soon as the heart button was pressed.

File: source/views/building-hours/detail/index.js (hunk 3 above)
Comment (hawkrives) [reply source/views/building-hours/detail/index.js]: Essentially, then, we avoid mutating the array here entirely and call onSaveBuildingFavorites or whatever it's called (sorry, reviewing from my phone atm) instead. Redux will cause a re-render or the screen when the list of favorited buildings changes, so we'll know that the building is favorited because it suddenly will be (if that makes sense; I can try to explain better if you want me to.)

Comment (hawkrives) [reply source/views/building-hours/detail/index.js]: You'll also need to add the dispatch call / map thing in here, instead of in StatefulList, because that's how this component will have access to the function.

File: source/views/building-hours/stateful-list.js [hunk 7]
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
+  if (!state.buildings) {
+    return {favoriteBuildings: []}
+  }
+
+  return {
+    favoriteBuildings: state.buildings.favoriteBuildings,
+  }
+}
+
+function mapDispatch(dispatch): ReduxDispatchProps {
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: This one will need to move in to the Detail screen, because we'll be calling onSaveFavorites from there.

File: source/views/components/nav-buttons/favorite.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+  favorited: boolean,
+}
+
+type Props = {
+  onFavorite: () => any,
+  favorited: boolean,
+}
+
+export class FavoriteButton extends React.PureComponent<Props, State> {
+  state = {
```
Comment (hawkrives) [inline source/views/components/nav-buttons/favorite.js]: When you move the logic of saving the favorite into the redux file and make the "am I favorited" logic on the Detail screen be a (mathematical) function of the Redux state, this component won't need any internal state any longer; the lifecycle will go something like this:

- Person taps "favorite"
- Redux "action creator" onSaveFavorites is called
- Said function determines an array of actually favorited buildings (and saves it to disk)
- Created action is dispatched via the `connect`ed dispatch function
- Redux state updates with new array of favorites building names
- Redux state update causes React to re-render the components that have connected to the store
- StatefulList gets the new set of building names and groups them again
- DetailView gets the new set of building names and figures out if it's favorited
- If it is, it sets the "favorited" prop of FavoriteButton to "true" or "false", and react calls FavoriteButton's render function again with the new props (otherwise nothing happens)

At that point, there's no reason to keep the separate state in this button, so it becomes a much smaller component. (Not that it's not already small.)

If you try that and the button doesn't seem to be working, please let me know - the router that handles new screens might be interfering with the props getting passed right (it may not be passing the new ones, since navigate is only called once and we haven't changed screens). Or it may not! I'm not sure.

File: source/views/building-hours/stateful-list.js [hunk 9]
```diff
@@ -92,24 +117,51 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
+  onUpdate = () => {
+    this.fetchData()
+    this.props.onSaveFavorites(this.props.favoriteBuildings)
+  }
+
   render() {
     if (this.state.error) {
       return <NoticeView text={`Error: ${this.state.error.message}`} />
     }
-
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: Please put this space back in :)

File: source/views/building-hours/stateful-list.js [hunk 10]
```diff
@@ -48,16 +74,15 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
     // now: moment.tz('Wed 7:25pm', 'ddd h:mma', null, CENTRAL_TZ),
     now: moment.tz(CENTRAL_TZ),
-    buildings: groupBuildings(defaultData.data),
+    buildings: loadAllBuildings(defaultData.data, this.props.favoriteBuildings),
     intervalId: 0,
   }
 
   componentWillMount() {
     this.fetchData()
-
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: Please put this space back :)

File: source/views/building-hours/detail/index.js [hunk 11]
```diff
@@ -37,8 +65,11 @@ export class BuildingHoursDetailView extends React.PureComponent<Props, State> {
     this.setState({intervalId: setInterval(this.updateTime, 1000)})
   }
 
+  componentDidMount() {}
```
Comment (hawkrives) [inline source/views/building-hours/detail/index.js]: You can probably remove this method; I don't think you'll need it here.

File: ios/AllAboutOlaf/Info.plist [hunk 12]
```diff
@@ -76,7 +76,7 @@
 	<key>NSCalendarsUsageDescription</key>
 	<string>We use your calendar to add events to your calendar so that you remember what you wanted to attend.</string>
 	<key>NSLocationWhenInUseUsageDescription</key>
-	<string>We only use your location when you would expect us to, and never when you aren&apos;t using the app.</string>
+	<string>We only use your location when you would expect us to, and never when you aren't using the app.</string>
```
Comment (hawkrives) [inline ios/AllAboutOlaf/Info.plist]: Please revert this (Xcode changes it, and something else changes it back… it's a struggle)

**Review by hawkrives — CHANGES_REQUESTED** [review]
@hannesmcman Wow, thanks!

I left several comments and one novel this time around; I think you can push this through in the next few days (if you need help or have questions with it, please let us know; we know that the docs for how this ball of mud works essentially don't exist).

But overall, good work!

**Comment by hawkrives** [comment]
@hannesmcman Regarding "npm run check" – if you run `yarn` in this project (after installing it from yarnpkg.com), does the "check" script work?

File: source/views/building-hours/stateful-list.js (hunk 5 above)
Comment (drewvolz) [reply source/views/building-hours/stateful-list.js]: I think he was going off of the comment that says "ten seconds" so let's change the comment if we're going to leave the code at one second.

Comment (hawkrives) [reply source/views/building-hours/stateful-list.js]: Yes, we should change the comment.

File: source/views/components/nav-buttons/favorite.js (hunk 8 above)
Comment (hannesmcman) [reply source/views/components/nav-buttons/favorite.js]: Ok, I have implemented the changes you asked for, but I am unsure how to get the navigationOptions in the detail view to refresh when the view is re-rendered by redux upon the state change. The only way I can get the button to update properly is if I pass navigation to the onSaveFavorites and saveFavoriteBuildings functions, respectively and then call navigation.setParams({favoriteBuildings: favoriteBuildings}) within saveFavoriteBuildings in buildings.js...is there a way to get the navigationOptions to automatically refresh when the redux state changes?

Comment (hannesmcman) [reply source/views/components/nav-buttons/favorite.js]: @hawkrives Regarding "npm run check"..running yarn does not make the "check" script work..

File: source/views/building-hours/stateful-list.js [hunk 13]
```diff
@@ -27,7 +29,31 @@ const groupBuildings = (buildings: BuildingType[]) => {
...
 
-type Props = TopLevelViewPropsType
+const loadAllBuildings = (
+  buildings: BuildingType[],
+  favoriteBuildings: BuildingType[],
+) => {
+  let allBuildings = Array()
+  allBuildings.push({
+    title: 'Favorites',
+    data: favoriteBuildings,
```
Comment (drewvolz) [inline source/views/building-hours/stateful-list.js]: I'd like to see logic incorporated that restricted the 'Favorites' section from appearing if there are no favorites, if you don't mind. If you don't want to implement this, I'd be happy to work on it.

File: source/views/components/nav-buttons/favorite.js (hunk 8 above)
Comment (hawkrives) [reply source/views/components/nav-buttons/favorite.js]: Sigh… that's what I was afraid of, that the toolbar buttons wouldn't be able to subscribe to redux. 

I'll think about this and probably get back to you tomorrow … I'm thinking that it might be possible / simple to connect to redux on the FavoriteButton itself, and pass the building name to the button, then use the buildings name in mapStateToProps within the button's redux connection to determine if the building is connected or not. 

I'll try to work up something for that approach tomorrow.

Comment (hannesmcman) [reply source/views/components/nav-buttons/favorite.js]: Sounds good. I just pushed the latest version that implemented most of the changes you requested.

File: source/views/building-hours/stateful-list.js (hunk 13 above)
Comment (hannesmcman) [reply source/views/building-hours/stateful-list.js]: https://github.com/StoDevX/AAO-React-Native/pull/2003/commits/3cdeb6c37bed1768cc0ad1a0e5f0ed6d74cf52d8 
I did it the easy way in this commit, but if you think there's a better way you should go for it. 👍

File: source/views/building-hours/detail/index.js [hunk 14]
```diff
@@ -98,10 +107,22 @@ export class BuildingHoursDetailView extends React.PureComponent<Props, State> {
 
 function mapStateToProps(state: ReduxState): ReduxStateProps {
   return {
-    favoriteBuildings: state.buildings ? state.buildings.favoriteBuildings : [],
+    favoriteBuildings: state.buildings.favoriteBuildings
```
Comment (hawkrives) [inline source/views/building-hours/detail/index.js]: We know that if state.buildings exists, then favoriteBuildings will too; it's the buildings key itself that may not exist (this line was correct before [unless something was yelling about it in which c add I am probably wrong])

File: source/views/building-hours/stateful-list.js [hunk 15]
```diff
@@ -92,19 +117,21 @@ export class BuildingHoursView extends React.Component<Props, State> {
...
     }
     this.setState(() => ({
-      buildings: groupBuildings(buildings),
+      buildings: groupBuildings(buildings, this.props.favoriteBuildings),
       now: moment.tz(CENTRAL_TZ),
     }))
   }
 
   render() {
+    this.fetchData()
```
Comment (hawkrives) [inline source/views/building-hours/stateful-list.js]: Why did you add this call here?

Comment (hannesmcman) [reply source/views/building-hours/stateful-list.js]: Without it, the favorites are not added in the re-render of stateful-list, because groupBuildings is not called.

**Comment by hawkrives** [comment]
@hannesmcman I pulled your branch to test my thoughts about connecting the toolbar button to Redux; it seems to work. You can see my commits on the https://github.com/StoDevX/AAO-React-Native/tree/add-favorite-buildings-notes branch. 

I think it works, but I didn't have time to actually build the app on your branch to check before my plane boards. 

I tried to avoid extraneous changes on my branch, but there might be some (sorry!). 

I left more details in most of the commit messages on my branch; I recommend reading them. 

Thanks again for spearheading this! And, wow, I'm impressed that you just dived right in to not only the app, but also Redux and state management!

---

"yarn check" just runs four other scripts, so you can try running those individually yourself to replicate the effect of the whole command and omit whichever one(s) are failing for you.

**Comment by hannesmcman** [comment]
@hawkrives Thanks for the help! I pulled your branch and it didn't seem to be working on my iOS emulator..I made a few comments where I think the issues may be, but it looks like aside from those issues your branch will be the one to use! Is your plan to just merge that branch into master when it's ready?

**Comment by hawkrives** [comment]
No, with your permission I'd like to push my commits into your branch so they show up on this PR.

I replied to both your comments, and one of them should fix the issue of it not working I hope.

**Comment by hannesmcman** [comment]
Sounds good. Go for it. Thanks!

**Comment by hawkrives** [comment]
@hannesmcman I merged my changes into your branch here. (I also reverted the changes to the xcode project, so you'll have to tweak that again to build it on your machine, probably.)

I'm going to actually run the app now and see what it looks like!

**Review by hawkrives — APPROVED** [review]
@hannesmcman I think this is ready to merge and ship as part of 2.5! Great work on spearheading this!

**Comment by hawkrives** [comment]
@rye, @drewvolz: I believe that this will pass when it's merged into `master`. I will babysit the build until it passes.

**Comment by hawkrives** [comment]
… I suppose I should have waited for a second sign-off. Sorry!

**Comment by rye** [comment]
Well. I'll take a look at it after I finish snowblowing. We can always fix stuff later.

@hawkrives, can you change the version to 2.5.0-pre?

**Comment by hawkrives** [comment]
@rye Done in #2011
